### PR TITLE
v1.8: Add new embedder sources

### LINKED ISSUE #2795: v1.8: Generic REST embedder
v1.8 introduces a generic embedder option compatible with any embedders offering a RESTful interface.

## Tasks

- [ ] Add new section to [Vector search page](https://www.meilisearch.com/docs/learn/experimental/vector_search#generate-vector-embeddings)

## References
- engine issue: https://github.com/meilisearch/meilisearch/issues/4531
- PRD: https://meilisearch.notion.site/v1-8-AI-search-API-usage-135552d6e85a4a52bc7109be82aeca42#e6f58c3b742c4effb4ddc625ce12ee16
- SME: @dureuill

### PR DESCRIPTION
**Pull request by guimachiavelli**
Closes #2795, #2796

### CODE REVIEW

File: learn/experimental/vector_search.mdx [hunk 1]
```diff
@@ -102,14 +103,84 @@ curl \
...
+
+`model` is a mandatory field indicating a compatible model.
+
+`documentTemplate` is an optional field. Use it to customize the data you send to the embedder. It is highly recommended you configure a custom template for your documents.
+</Tabs.Content>
+</Tabs.Container>
 
 <Capsule intent="tip" title="`documentTemplate` usage">
-`documentTemplate` must be a [Liquid template](https://shopify.github.io/liquid/). Use `{{ doc.attribute }}` to access document field values. Meilisearch also exposes a `{{ fields }}` array containing one object per document field, which you may access with `{{ field.name }}` and `{{ field.value }}`.
+`documentTemplate` must be a [Liquid template](https://shopify.github.io/liquid/). Use `{{ doc.attribute }}` to access the `attribute` field value of your documents. Any field you refer to in this way must exist in all documents or an error will be raised at indexing time.
```
Comment (guimachiavelli) [inline learn/experimental/vector_search.mdx:178]:
```suggestion
`documentTemplate` must be a [Liquid template](https://shopify.github.io/liquid/). Use `{{ doc.attribute }}` to access the `attribute` field value of your documents. Meilisearch also exposes a `{{ fields }}` array containing one object per document field, which you may access with `{{ field.name }}` and `{{ field.value }}`.

Any field you refer to in this way must exist in all documents or an error will be raised at indexing time.
```

**Comment by macraig** [comment]
@guimachiavelli I think we might be missing the `distributionShift` setting? ([API link](https://www.notion.so/meilisearch/v1-8-AI-search-API-usage-135552d6e85a4a52bc7109be82aeca42?pvs=4#d652adc0890445658aaf36352dbc8802))

**Comment by guimachiavelli** [comment]
Thanks for bringing that up, @macraig. `distributionShift` was bundled with other hybrid search changes and I missed it when planning v1.8 updates. I'll open a separate PR to cover that.

File: learn/experimental/vector_search.mdx [hunk 2]
```diff
@@ -37,14 +37,21 @@ To ensure proper scaling of Meilisearch Cloud's latest AI-powered search offerin
...
+Embedders generate vector data from your documents. Meilisearch natively supports the following embedders:
 
-It is also possible to use custom embedders. In this case, you must generate the embeddings manually and include them as a field in your documents.
+- [OpenAI](https://openai.com/)
+- [Hugging Face](https://huggingface.co/)
+- [Ollama](https://ollama.com/)
+- any embedder with a REST interface
 
-#### Generate auto-embeddings with OpenAI 
+It is also possible to supply custom embeddings. In this case, you must generate the embeddings manually and include them as a field in your documents.
```
Comment (Kerollmops) [inline learn/experimental/vector_search.mdx:49]: Do you want to specify the `_vectors` field or link to somewhere that explains this, maybe?

File: learn/experimental/vector_search.mdx [hunk 3]
```diff
@@ -102,14 +103,86 @@ curl \
...
+  -X PATCH 'http://localhost:7700/indexes/movies/settings' \
+  -H 'Content-Type: application/json' \
+  --data-binary '{
+    "embedders": {
+      "default": {
+        "source": "ollama",
+        "url": "http://localhost:11434/api/embeddings",
+        "apiKey": "OLLAMA_API_KEY",
+        "model": "MODEL_NAME",
+        "documentTemplate": "A document titled '{{doc.title}}' whose description starts with {{doc.overview|truncatewords: 20}}"
```
Comment (Kerollmops) [inline learn/experimental/vector_search.mdx:122]:
```suggestion
        "documentTemplate": "A document titled {{doc.title}} whose description starts with {{doc.overview|truncatewords: 20}}"
```

BTW, it is not the only code sample part I notice with single quotes that "collide" with the shell single quotes. I advise you to remove them in those examples (it's a nightmare to put single quotes in a single quoted string in a shell expression) 😢

File: learn/experimental/vector_search.mdx [hunk 4]
```diff
@@ -102,14 +103,86 @@ curl \
...
+      }
+    }
+  }'
+```
+
+`url` indicates the embedder address. If not specified, Meilisearch will try to fetch the `MEILI_OLLAMA_URL` environment variable.
+
+`apiKey` is an optional field indicating your Ollama API key, if you have configured it.
+
+`model` is a mandatory field indicating a compatible model.
```
Comment (Kerollmops) [inline learn/experimental/vector_search.mdx:132]: You can find them [at this address](https://ollama.com/library).

File: learn/experimental/vector_search.mdx [hunk 5]
```diff
@@ -102,14 +103,86 @@ curl \
...
+        "dimensions": 512,
+        "documentTemplate": "A document titled '{{doc.title}}' whose description starts with {{doc.overview|truncatewords: 20}}",
+        "inputField": ["data", "text"],
+        "inputType": "text",
+        "query": {
+          "model": "MODEL_NAME",
+          "dimensions": 512
+        },
+        "pathToEmbeddings": ["data"],
+        "embeddingObject": ["embedding"]
```
Comment (Kerollmops) [inline learn/experimental/vector_search.mdx:161]: Do we plan to define what `data`, `text` or other field means?

**Review by Kerollmops — CHANGES_REQUESTED** [review]

File: learn/experimental/vector_search.mdx (hunk 5 above)
Comment (guimachiavelli) [reply learn/experimental/vector_search.mdx:161]: If you think it's fundamental to explain it, I'm happy to add definitions. Two things, though:

1. I couldn't find them in the reference document, so I'd need someone to explain them to me (or point me to a place where I can find that information)
2. This document is already super bloated. Really, it's a huge bag of "everything related to vector search", so I'm trying not to add more information than strictly necessary. My plan is to soon break this down in more focussed articles

Comment (guimachiavelli) [reply learn/experimental/vector_search.mdx:161]: I have discussed this with @macraig and we decided to leave this PR without any definitions for the moment. Once both louis and myself are back from our holidays, I'll create a new PR adding more information.

**Pull request merged**